websites/Y/YouTube Music/presence.ts

```diff
@@ -2,90 +2,263 @@ const presence = new Presence({
 	clientId: "463151177836658699",
 });
 
-let prevMetadata: {
-	title: string;
-	startedAt: number;
-};
+let prevTitleAuthor = "",
+	presenceData: PresenceData,
+	mediaTimestamps: [number, number],
+	oldPath: string,
+	startTimestamp: number,
+	videoListenerAttached = false;
 
 presence.on("UpdateData", async () => {
-	const title = navigator.mediaSession.metadata?.title,
-		video = document.querySelector<HTMLVideoElement>(".video-stream"),
-		progressBar = document.querySelector<HTMLElement>("#progress-bar"),
-		repeatMode = document
-			.querySelector('ytmusic-player-bar[slot="player-bar"]')
-			.getAttribute("repeat-Mode_"),
-		[buttons, timestamps, cover] = await Promise.all([
+	const [
+			showButtons,
+			showTimestamps,
+			showCover,
+			hidePaused,
+			showBrowsing,
+			privacyMode,
+		] = await Promise.all([
 			presence.getSetting<boolean>("buttons"),
 			presence.getSetting<boolean>("timestamps"),
 			presence.getSetting<boolean>("cover"),
-		]);
-	if (title && !isNaN(video.duration)) {
-		const endTimestamp =
-				Date.now() / 1000 +
-				Number(progressBar.getAttribute("aria-valuemax")) -
-				Number(progressBar.getAttribute("value")),
-			watchID = document
-				.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
-				.href.match(/v=([^&#]{5,})/)[1],
-			presenceData: PresenceData = {
-				details: title,
-				state: [
-					navigator.mediaSession.metadata.artist,
-					navigator.mediaSession.metadata.album,
-				]
-					.filter(Boolean)
-					.join(" - "),
-				largeImageKey: cover
-					? navigator.mediaSession.metadata.artwork[0].src
-					: "ytm_lg",
-				smallImageKey: video.paused
+			presence.getSetting<boolean>("hidePaused"),
+			presence.getSetting<boolean>("browsing"),
+			presence.getSetting<boolean>("privacy"),
+		]),
+		{ mediaSession } = navigator,
+		watchID = document
+			.querySelector<HTMLAnchorElement>("a.ytp-title-link.yt-uix-sessionlink")
+			.href.match(/v=([^&#]{5,})/)?.[1],
+		repeatMode = document
+			.querySelector('ytmusic-player-bar[slot="player-bar"]')
+			.getAttribute("repeat-Mode_"),
+		videoElement =
+			document.querySelector<HTMLVideoElement>("video.video-stream");
+
+	if (videoElement) {
+		if (!videoListenerAttached) {
+			const videoListener = () => {
+				mediaTimestamps = presence.getTimestampsfromMedia(videoElement);
+				//* Don't ask me why the above function doesn't floor the end timestamp
+				mediaTimestamps[1] = Math.floor(mediaTimestamps[1]);
+			};
+
+			//* If video scrobbled, update timestamps
+			videoElement.addEventListener("seeked", videoListener);
+			//* If video resumes playing, update timestamps
+			videoElement.addEventListener("play", videoListener);
+
+			videoListenerAttached = true;
+		}
+		//* Element got removed from the DOM (eg, song with song/video switch)
```

Comment:
Not needed

Reply:
How is this not needed?

Reply:
It explains exactly what the else does in this statement.

Reply:
I would keep that comment, because if we look at the code in let's say a year someone is going to be like "It's always defined" but in reality it is not if you switch